### PR TITLE
planner: add HashCode method for LogicalPlan & use PlanHash for GroupExpr's fingerPrint

### PR DESCRIPTION
**Pull request by francis0407**
### What problem does this PR solve? 
1. Add `HashCode()` method for LogicalPlan.
2. Implement `HashCode()` for some simple plans: Projection, Selection, Limit, TableDual.
3. Change the GroupExpr's FingerPrint to plan's HashCode.

### What is changed and how it works?

`HashCode()` uses a precise hash method to encode the LogicalPlan. For example, `LogicalProjection` uses a format `PlanType + ExprNum + [Exprs...]`.

### Check List 

Tests 

 - Unit test

Code changes

 - Has interface methods change

Side effects

 - Possible performance regression
 - Increased code complexity

### CODE REVIEW

**Comment by francis0407** [comment]
> Though it's very strange, but... we need to take the `SelectBlockOffset` into consideration?

Not quite sure if it is necessary. cc @lamxTyler

**Comment by lamxTyler** [comment]
Why? Plans in different block would have different children?

File: planner/core/hashcode.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	var buf [4]byte
+	binary.BigEndian.PutUint32(buf[:], uint32(value))
+	return append(result, buf[:]...)
+}
+
+// HashCode implements LogicalPlan interface.
+func (p *baseLogicalPlan) HashCode() []byte {
+	// PlanType + PlanID
+	// We use PlanID for the default hash, so if two plans do not have
+	// the same id, the hash value will never be the same.
```
Comment (lamxTyler) [inline planner/core/hashcode.go]: Then why add the plan type?

File: planner/core/hashcode.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+// HashCode implements LogicalPlan interface.
+func (p *LogicalProjection) HashCode() []byte {
+	// PlanType + ExprNum + [Exprs]
+	result := make([]byte, 0, 8)
+	result = encodeIntAsUint32(result, plancodec.TypeStringToPhysicalID(p.tp))
+	result = encodeIntAsUint32(result, len(p.Exprs))
+	for _, expr := range p.Exprs {
+		exprHashCode := expr.HashCode(p.ctx.GetSessionVars().StmtCtx)
```
Comment (lamxTyler) [inline planner/core/hashcode.go]: Why encode the stmtctx?

Comment (francis0407) [reply planner/core/hashcode.go]: This is `Expression.HashCode(ctx)`, it returns the hash code of an expression.

File: planner/core/hashcode.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	result := make([]byte, 0, 8)
+	result = encodeIntAsUint32(result, plancodec.TypeStringToPhysicalID(p.tp))
+	result = encodeIntAsUint32(result, p.RowCount)
+	return result
+}
+
+// HashCode implements LogicalPlan interface.
+func (p *LogicalSelection) HashCode() []byte {
+	// PlanType + ConditionNum + [Conditions]
+	result := make([]byte, 0, 8)
```
Comment (lzmhhh123) [inline planner/core/hashcode.go]: We can malloc more space for the result which append the expression. In the selection case, `8 + len(p.Condition) * &expression.Column{}.HashCode()`.

File: planner/core/hashcode.go (hunk 1 above)
Comment (francis0407) [reply planner/core/hashcode.go]: addressed

File: planner/core/hashcode.go (hunk 3 above)
Comment (francis0407) [reply planner/core/hashcode.go]: addressed

File: planner/core/hashcode.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+func (p *LogicalProjection) HashCode() []byte {
+	// PlanType + ExprNum + [Exprs]
+	// Expressions are commonly `Column`s, whose hashcode has the length 9, so
+	// we pre-alloc 10 bytes for each expr's hashcode.
+	result := make([]byte, 0, 8+len(p.Exprs)*10)
+	result = encodeIntAsUint32(result, plancodec.TypeStringToPhysicalID(p.tp))
+	result = encodeIntAsUint32(result, len(p.Exprs))
+	for _, expr := range p.Exprs {
+		exprHashCode := expr.HashCode(p.ctx.GetSessionVars().StmtCtx)
+		println(len(exprHashCode))
```
Comment (lzmhhh123) [inline planner/core/hashcode.go]: What's the line for?

Comment (francis0407) [reply planner/core/hashcode.go]: removed

**Comment by winoros** [comment]
@lamxTyler The ideal hashcode/fingerprint should mark two different `DataSource(t)->Selection(a>1)` the same one if the the things like output column are also the same.
So i think we may need to take the offset into consideration?

**Comment by lamxTyler** [comment]
@winoros Yes, you are right.

**Pull request merged**